File: swift/Sources/TransactionLog/TransactionLogCodec.swift
import Foundation
import SdJwt

public struct TransactionLogCodecError: Error, CustomStringConvertible {
    public let description: String
    init(_ description: String) { self.description = description }
}

/// JSON serialization for `TransactionLogEntry` so a persistent `TransactionLogStore` can store
/// entries as text. Byte fields are base64url; the shape mirrors the Kotlin codec for a stable,
/// cross-platform on-disk format.
public enum TransactionLogCodec {

    public static func encode(_ entry: TransactionLogEntry) -> String { toJson(entry).serialize() }

    public static func decode(_ text: String) throws -> TransactionLogEntry {
        try fromJson(try JsonValue.parse(text))
    }

    static func toJson(_ e: TransactionLogEntry) -> JsonValue {
        var o: [(String, JsonValue)] = [
            ("id", .str(e.id)),
            ("timestamp", .numInt(e.timestamp)),
            ("type", .str(e.type.rawValue)),
            ("status", .str(e.status.rawValue)),
        ]
        if let rp = e.relyingParty { o.append(("relyingParty", rpJson(rp))) }
        if let issuer = e.issuer { o.append(("issuer", .str(issuer))) }
        o.append(("documents", .arr(e.documents.map(docJson))))
        if let error = e.error { o.append(("error", .str(error))) }
        if let raw = e.rawRequest { o.append(("rawRequest", .str(Base64Url.encode(raw)))) }
        if let raw = e.rawResponse { o.append(("rawResponse", .str(Base64Url.encode(raw)))) }
        if let transport = e.transport { o.append(("transport", .str(transport.rawValue))) }
        if let issuerName = e.issuerName { o.append(("issuerName", .str(issuerName))) }
        if let issuerRegistered = e.issuerRegistered { o.append(("issuerRegistered", .bool(issuerRegistered))) }
        return .obj(o)
    }

    static func fromJson(_ json: JsonValue) throws -> TransactionLogEntry {
        guard case .obj = json else { throw TransactionLogCodecError("entry must be an object") }
        guard case let .str(id)? = json["id"], case let .numInt(ts)? = json["timestamp"],
              case let .str(typeRaw)? = json["type"], let type = TransactionType(rawValue: typeRaw),
              case let .str(statusRaw)? = json["status"], let status = TransactionStatus(rawValue: statusRaw)
        else { throw TransactionLogCodecError("entry missing required fields") }

        var rp: RelyingParty?
        if let rpJson = json["relyingParty"], case .obj = rpJson { rp = rpFromJson(rpJson) }
        var issuer: String?
        if case let .str(s)? = json["issuer"] { issuer = s }
        var documents: [LoggedDocument] = []
        if case let .arr(items)? = json["documents"] { documents = items.compactMap { if case .obj = $0 { return docFromJson($0) }; return nil } }
        var error: String?
        if case let .str(s)? = json["error"] { error = s }
        var rawRequest: [UInt8]?
        if case let .str(s)? = json["rawRequest"] { rawRequest = try Base64Url.decode(s) }
        var rawResponse: [UInt8]?
        if case let .str(s)? = json["rawResponse"] { rawResponse = try Base64Url.decode(s) }
        var transport: TransactionTransport?
        if case let .str(s)? = json["transport"] { transport = TransactionTransport(rawValue: s) }
        var issuerName: String?
        if case let .str(s)? = json["issuerName"] { issuerName = s }
        var issuerRegistered: Bool?
        if case let .bool(b)? = json["issuerRegistered"] { issuerRegistered = b }

        return TransactionLogEntry(id: id, timestamp: ts, type: type, status: status, relyingParty: rp,
                                   issuer: issuer, documents: documents, error: error,
                                   rawRequest: rawRequest, rawResponse: rawResponse,
                                   transport: transport, issuerName: issuerName, issuerRegistered: issuerRegistered)
    }

    private static func rpJson(_ rp: RelyingParty) -> JsonValue {
        var o: [(String, JsonValue)] = [("id", .str(rp.id))]
        if let name = rp.name { o.append(("name", .str(name))) }
        o.append(("trusted", .bool(rp.trusted)))
        if !rp.certificateChainDer.isEmpty {
            o.append(("certificateChain", .arr(rp.certificateChainDer.map { .str(Base64Url.encode($0)) })))
        }
        if let clientIdScheme = rp.clientIdScheme { o.append(("clientIdScheme", .str(clientIdScheme))) }
        if let subject = rp.subject { o.append(("subject", .str(subject))) }
        if !rp.entitlements.isEmpty { o.append(("entitlements", .arr(rp.entitlements.map { .str($0) }))) }
        if !rp.purpose.isEmpty { o.append(("purpose", .arr(rp.purpose.map(textJson)))) }
        if let intermediaryName = rp.intermediaryName { o.append(("intermediaryName", .str(intermediaryName))) }
        if let intermediarySub = rp.intermediarySub { o.append(("intermediarySub", .str(intermediarySub))) }
        if let attested = rp.attested { o.append(("attested", .bool(attested))) }
        if let statusValid = rp.statusValid { o.append(("statusValid", .bool(statusValid))) }
        return .obj(o)
    }

    private static func rpFromJson(_ json: JsonValue) -> RelyingParty {
        var id = ""; if case let .str(s)? = json["id"] { id = s }
        var name: String?; if case let .str(s)? = json["name"] { name = s }
        var trusted = false; if case let .bool(b)? = json["trusted"] { trusted = b }
        var chain: [[UInt8]] = []
        if case let .arr(items)? = json["certificateChain"] { chain = items.compactMap { if case let .str(s) = $0 { return try? Base64Url.decode(s) }; return nil } }
        var clientIdScheme: String?; if case let .str(s)? = json["clientIdScheme"] { clientIdScheme = s }
        var subject: String?; if case let .str(s)? = json["subject"] { subject = s }
        var entitlements: [String] = []
        if case let .arr(items)? = json["entitlements"] { entitlements = items.compactMap { if case let .str(s) = $0 { return s }; return nil } }
        var purpose: [LocalizedText] = []
        if case let .arr(items)? = json["purpose"] { purpose = items.compactMap { if case .obj = $0 { return textFromJson($0) }; return nil } }
        var intermediaryName: String?; if case let .str(s)? = json["intermediaryName"] { intermediaryName = s }
        var intermediarySub: String?; if case let .str(s)? = json["intermediarySub"] { intermediarySub = s }
        var attested: Bool?; if case let .bool(b)? = json["attested"] { attested = b }
        var statusValid: Bool?; if case let .bool(b)? = json["statusValid"] { statusValid = b }
        return RelyingParty(id: id, name: name, trusted: trusted, certificateChainDer: chain,
                            clientIdScheme: clientIdScheme, subject: subject, entitlements: entitlements,
                            purpose: purpose, intermediaryName: intermediaryName, intermediarySub: intermediarySub,
                            attested: attested, statusValid: statusValid)
    }

    private static func textJson(_ t: LocalizedText) -> JsonValue {
        .obj([("lang", .str(t.lang)), ("value", .str(t.value))])
    }

    private static func textFromJson(_ json: JsonValue) -> LocalizedText {
        var lang = ""; if case let .str(s)? = json["lang"] { lang = s }
        var value = ""; if case let .str(s)? = json["value"] { value = s }
        return LocalizedText(lang: lang, value: value)
    }

    private static func docJson(_ d: LoggedDocument) -> JsonValue {
        var o: [(String, JsonValue)] = [("format", .str(d.format))]
        if let type = d.type { o.append(("type", .str(type))) }
        if let queryId = d.queryId { o.append(("queryId", .str(queryId))) }
        o.append(("claims", .arr(d.claims.map(claimJson))))
        return .obj(o)
    }

    private static func docFromJson(_ json: JsonValue) -> LoggedDocument {
        var format = ""; if case let .str(s)? = json["format"] { format = s }
        var type: String?; if case let .str(s)? = json["type"] { type = s }
        var queryId: String?; if case let .str(s)? = json["queryId"] { queryId = s }
        var claims: [LoggedClaim] = []
        if case let .arr(items)? = json["claims"] { claims = items.compactMap { if case .obj = $0 { return claimFromJson($0) }; return nil } }
        return LoggedDocument(format: format, type: type, queryId: queryId, claims: claims)
    }

    private static func claimJson(_ c: LoggedClaim) -> JsonValue {
        var o: [(String, JsonValue)] = [("path", .arr(c.path.map { .str($0) }))]
        if let value = c.value { o.append(("value", .str(value))) }
        return .obj(o)
    }

    private static func claimFromJson(_ json: JsonValue) -> LoggedClaim {
        var path: [String] = []
        if case let .arr(items)? = json["path"] { path = items.compactMap { if case let .str(s) = $0 { return s }; return nil } }
        var value: String?; if case let .str(s)? = json["value"] { value = s }
        return LoggedClaim(path: path, value: value)
    }
}

File: kotlin/txlog/src/main/kotlin/com/hopae/eudi/wallet/txlog/TransactionLogCodec.kt
package com.hopae.eudi.wallet.txlog

import com.hopae.eudi.wallet.sdjwt.Base64Url
import com.hopae.eudi.wallet.sdjwt.JsonValue

/**
 * JSON serialization for [TransactionLogEntry] so a persistent [TransactionLogStore] can store
 * entries as text. Byte fields are base64url; the shape mirrors the Swift codec for a stable,
 * cross-platform on-disk format.
 */
object TransactionLogCodec {

    fun encode(entry: TransactionLogEntry): String = toJson(entry).serialize()

    fun decode(text: String): TransactionLogEntry =
        fromJson(JsonValue.parse(text) as? JsonValue.Obj ?: error("transaction log entry must be a JSON object"))

    fun toJson(e: TransactionLogEntry): JsonValue.Obj = JsonValue.Obj(
        buildList {
            add("id" to JsonValue.Str(e.id))
            add("timestamp" to JsonValue.NumInt(e.timestamp))
            add("type" to JsonValue.Str(e.type.name))
            add("status" to JsonValue.Str(e.status.name))
            e.relyingParty?.let { add("relyingParty" to rpJson(it)) }
            e.issuer?.let { add("issuer" to JsonValue.Str(it)) }
            add("documents" to JsonValue.Arr(e.documents.map { docJson(it) }))
            e.error?.let { add("error" to JsonValue.Str(it)) }
            e.rawRequest?.let { add("rawRequest" to JsonValue.Str(Base64Url.encode(it))) }
            e.rawResponse?.let { add("rawResponse" to JsonValue.Str(Base64Url.encode(it))) }
            e.transport?.let { add("transport" to JsonValue.Str(it.name)) }
            e.issuerName?.let { add("issuerName" to JsonValue.Str(it)) }
            e.issuerRegistered?.let { add("issuerRegistered" to JsonValue.Bool(it)) }
        }
    )

    fun fromJson(o: JsonValue.Obj): TransactionLogEntry = TransactionLogEntry(
        id = o.str("id") ?: error("missing id"),
        timestamp = o.num("timestamp") ?: error("missing timestamp"),
        type = TransactionType.valueOf(o.str("type") ?: error("missing type")),
        status = TransactionStatus.valueOf(o.str("status") ?: error("missing status")),
        relyingParty = (o["relyingParty"] as? JsonValue.Obj)?.let { rpFromJson(it) },
        issuer = o.str("issuer"),
        documents = (o["documents"] as? JsonValue.Arr)?.items?.mapNotNull { (it as? JsonValue.Obj)?.let(::docFromJson) } ?: emptyList(),
        error = o.str("error"),
        rawRequest = o.str("rawRequest")?.let { Base64Url.decode(it) },
        rawResponse = o.str("rawResponse")?.let { Base64Url.decode(it) },
        transport = o.str("transport")?.let { runCatching { TransactionTransport.valueOf(it) }.getOrNull() },
        issuerName = o.str("issuerName"),
        issuerRegistered = (o["issuerRegistered"] as? JsonValue.Bool)?.value,
    )

    private fun rpJson(rp: RelyingParty) = JsonValue.Obj(
        buildList {
            add("id" to JsonValue.Str(rp.id))
            rp.name?.let { add("name" to JsonValue.Str(it)) }
            add("trusted" to JsonValue.Bool(rp.trusted))
            if (rp.certificateChainDer.isNotEmpty()) {
                add("certificateChain" to JsonValue.Arr(rp.certificateChainDer.map { JsonValue.Str(Base64Url.encode(it)) }))
            }
            rp.clientIdScheme?.let { add("clientIdScheme" to JsonValue.Str(it)) }
            rp.subject?.let { add("subject" to JsonValue.Str(it)) }
            if (rp.entitlements.isNotEmpty()) add("entitlements" to JsonValue.Arr(rp.entitlements.map { JsonValue.Str(it) }))
            if (rp.purpose.isNotEmpty()) add("purpose" to JsonValue.Arr(rp.purpose.map { textJson(it) }))
            rp.intermediaryName?.let { add("intermediaryName" to JsonValue.Str(it)) }
            rp.intermediarySub?.let { add("intermediarySub" to JsonValue.Str(it)) }
            rp.attested?.let { add("attested" to JsonValue.Bool(it)) }
            rp.statusValid?.let { add("statusValid" to JsonValue.Bool(it)) }
        }
    )

    private fun rpFromJson(o: JsonValue.Obj) = RelyingParty(
        id = o.str("id") ?: error("relyingParty missing id"),
        name = o.str("name"),
        trusted = (o["trusted"] as? JsonValue.Bool)?.value ?: false,
        certificateChainDer = (o["certificateChain"] as? JsonValue.Arr)?.items?.mapNotNull { (it as? JsonValue.Str)?.value?.let(Base64Url::decode) } ?: emptyList(),
        clientIdScheme = o.str("clientIdScheme"),
        subject = o.str("subject"),
        entitlements = (o["entitlements"] as? JsonValue.Arr)?.items?.mapNotNull { (it as? JsonValue.Str)?.value } ?: emptyList(),
        purpose = (o["purpose"] as? JsonValue.Arr)?.items?.mapNotNull { (it as? JsonValue.Obj)?.let(::textFromJson) } ?: emptyList(),
        intermediaryName = o.str("intermediaryName"),
        intermediarySub = o.str("intermediarySub"),
        attested = (o["attested"] as? JsonValue.Bool)?.value,
        statusValid = (o["statusValid"] as? JsonValue.Bool)?.value,
    )

    private fun textJson(t: LocalizedText) = JsonValue.Obj(listOf("lang" to JsonValue.Str(t.lang), "value" to JsonValue.Str(t.value)))
    private fun textFromJson(o: JsonValue.Obj) = LocalizedText(o.str("lang") ?: "", o.str("value") ?: "")

    private fun docJson(d: LoggedDocument) = JsonValue.Obj(
        buildList {
            add("format" to JsonValue.Str(d.format))
            d.type?.let { add("type" to JsonValue.Str(it)) }
            d.queryId?.let { add("queryId" to JsonValue.Str(it)) }
            add("claims" to JsonValue.Arr(d.claims.map { claimJson(it) }))
        }
    )

    private fun docFromJson(o: JsonValue.Obj) = LoggedDocument(
        format = o.str("format") ?: error("document missing format"),
        type = o.str("type"),
        queryId = o.str("queryId"),
        claims = (o["claims"] as? JsonValue.Arr)?.items?.mapNotNull { (it as? JsonValue.Obj)?.let(::claimFromJson) } ?: emptyList(),
    )

    private fun claimJson(c: LoggedClaim) = JsonValue.Obj(
        buildList {
            add("path" to JsonValue.Arr(c.path.map { JsonValue.Str(it) }))
            c.value?.let { add("value" to JsonValue.Str(it)) }
        }
    )

    private fun claimFromJson(o: JsonValue.Obj) = LoggedClaim(
        path = (o["path"] as? JsonValue.Arr)?.items?.mapNotNull { (it as? JsonValue.Str)?.value } ?: emptyList(),
        value = o.str("value"),
    )

    private fun JsonValue.Obj.str(name: String): String? = (this[name] as? JsonValue.Str)?.value
    private fun JsonValue.Obj.num(name: String): Long? = (this[name] as? JsonValue.NumInt)?.value
}
